File: packages/ios/Sources/KakkaComponents/KakkaBadge.swift
import SwiftUI
import KakkaTokens

// MARK: - BadgeVariant
public enum BadgeVariant {
    case `default`
    case accent
    case success
    case warning
    case error
}

// MARK: - KakkaBadge
public struct KakkaBadge: View {
    let text: String
    let variant: BadgeVariant

    @Environment(\.kakkaTheme) private var theme

    public init(_ text: String, variant: BadgeVariant = .default) {
        self.text = text
        self.variant = variant
    }

    private var backgroundColor: Color {
        switch variant {
        case .default: return theme.surfaceColor
        case .accent:  return KakkaColors.accentPrimaryLight
        case .success: return KakkaColors.success.opacity(0.15)
        case .warning: return KakkaColors.warning.opacity(0.15)
        case .error:   return KakkaColors.error.opacity(0.15)
        }
    }

    private var foregroundColor: Color {
        switch variant {
        case .default: return theme.textSecondary
        case .accent:  return KakkaColors.accentPrimaryDark
        case .success: return KakkaColors.success
        case .warning: return KakkaColors.warning
        case .error:   return KakkaColors.error
        }
    }

    public var body: some View {
        Text(text)
            .font(KakkaFontSize.xs)
            .fontWeight(.medium)
            .foregroundColor(foregroundColor)
            .padding(.horizontal, KakkaSpacing.s2)
            .padding(.vertical, KakkaSpacing.s1)
            .background(backgroundColor)
            .clipShape(Capsule())
    }
}

// MARK: - Preview
#if DEBUG
#Preview {
    HStack(spacing: KakkaSpacing.s2) {
        KakkaBadge("Default", variant: .default)
        KakkaBadge("Accent", variant: .accent)
        KakkaBadge("Success", variant: .success)
        KakkaBadge("Warning", variant: .warning)
        KakkaBadge("Error", variant: .error)
    }
    .padding()
}
#endif

File: packages/android/kakka-components/src/main/kotlin/design/kakka/components/badge/KakkaBadge.kt
package design.kakka.components.badge

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import design.kakka.components.KakkaTheme
import design.kakka.tokens.KakkaColors
import design.kakka.tokens.KakkaRadius
import design.kakka.tokens.KakkaSpacing

enum class KakkaBadgeVariant { Default, Accent, Success, Warning, Error }

@Composable
fun KakkaBadge(
    text: String,
    modifier: Modifier = Modifier,
    variant: KakkaBadgeVariant = KakkaBadgeVariant.Default,
) {
    val (backgroundColor, contentColor) = when (variant) {
        KakkaBadgeVariant.Default -> Pair(
            MaterialTheme.colorScheme.primaryContainer,
            MaterialTheme.colorScheme.onSurface,
        )
        KakkaBadgeVariant.Accent -> Pair(
            KakkaColors.accentPrimaryDefault,
            KakkaColors.gray0,
        )
        KakkaBadgeVariant.Success -> Pair(
            KakkaColors.statusSuccess.copy(alpha = 0.15f),
            KakkaColors.statusSuccess,
        )
        KakkaBadgeVariant.Warning -> Pair(
            KakkaColors.statusWarning.copy(alpha = 0.15f),
            KakkaColors.statusWarning,
        )
        KakkaBadgeVariant.Error -> Pair(
            KakkaColors.statusError.copy(alpha = 0.15f),
            KakkaColors.statusError,
        )
    }

    Box(
        modifier = modifier
            .background(
                color = backgroundColor,
                shape = RoundedCornerShape(KakkaRadius.full),
            )
            .padding(horizontal = KakkaSpacing.s3, vertical = KakkaSpacing.s1),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = text,
            color = contentColor,
            style = MaterialTheme.typography.labelLarge,
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun KakkaBadgePreview() {
    KakkaTheme {
        KakkaBadge(text = "デフォルト")
    }
}
